Add withCookies to Java Result API

gmethvin opened the pull request:
Fixes #6047

cchantep commented on framework/src/play/src/main/scala/play/core/j/JavaResults.scala:
```diff
@@ -42,6 +42,19 @@ object JavaResultExtractor {
+  def withCookies(header: ResponseHeader, cookies: Array[JCookie]): ResponseHeader = {
+    if (cookies.isEmpty) {
+      header
+    } else {
```
```
else {
  val newCookies = JavaHelpers.cookiesToScalaCookies(java.util.Arrays.asList(cookies: _*))
  val headers = header.headers.get(HeaderNames.SET_COOKIE) match {
    case Some(existing) =>
      Cookies.mergeSetCookieHeader(existing, newCookies)
    case _ => newCookies // avoid merge logic
  }

  header.copy(headers = headers)
}
```

> Renaming `header` as `response` would be clearer for me there.

gmethvin replied:
@cchantep I can make this a bit more readable, but `mergeSetCookieHeader` also removes duplicate cookies and encodes the header, so I think it makes sense to use it even if all the cookies are all new here.

cchantep replied:
So maybe this deduplication/encoding logic should be moved outside of the merge one, to be usable in both case.

gmethvin replied:
I would prefer to improve the logic in `mergeSetCookieHeader` to better handle empty headers, since that's the method we actually use throughout our codebase. All of our APIs add cookies to a potentially non-empty set of cookies. To do that I updated `decodeSetCookieHeader` to return an empty list of cookies if the cookie header is empty. Let me know if that works for you.

marcospereira commented:
@gmethvin, could you please add some tests here?

schmitch commented:
Actually I'm not sure about the current way of handling that. 
Since it would diverge on how you would actually set things on the Java API.
Either we change **everything** to the .setMethod() or we use the same technique as you did in this PR.

The current way is actually:
https://www.playframework.com/documentation/2.5.x/JavaResponse#Setting-HTTP-response-headers
https://www.playframework.com/documentation/2.5.x/JavaSessionFlash

It uses the ThreadLocal Context to change the Response.
if we would merge this as is, then the way how Cookies could be added to the response would differ on how you would actually manipulate the response inside a Action.

gmethvin commented:
@schmitch The point of this PR is to make it possible to modify requests in a filter or in some other context that doesn't use Java Actions. I think we will want to remove the Java Http.Context at some point soon so there are not two ways of doing the same thing.

schmitch commented:
@gmethvin I just wanted to point out that this PR would actually also introduce that withCookies to the Action. And that it maybe could be confusing if the way how to manipulate a Response on Filters differs with the way how you would "actually" do it inside a Action.

Also with this change it's still *only* possible to change a Cookie inside the Filter, so maybe we should at least bring in the **most** useful methods, too. `withHeader`, `withHeaders` already exists. so I think it would actually also be good to manipulate the Session inside the Filter. Some people prefer to use that instead of raw Cookies. And maybe Flash Sessions. Also changing the Content-Type would be useful, too for gzip filters.

gmethvin commented:
@schmitch Yes, `withSession` and `withFlash` should also be provided, but that is a little bit more complicated to implement. There is actually some global state that we need to remove for Session and Flash as well.

I was actually thinking this change specifically should be backported to 2.5 first, since it's the minimum requirement to get any of the cookies working with filters. The session and flash cookie names can be obtained from the `HttpConfiguration`.

schmitch commented:
Don't forget the Content-Type change. If you manipulate it via Headers you won't change the Content-Type. You would get the following warning:
```
Content-Type set both in header (image/x-icon) and attached to entity (image/x-icon), ignoring content type from entity. To remove this warning, use Result.as(...) to set the content type, rather than setting the header manually.
```

And manipulating the Content-Type could be really useful in Filters (gzip for example).

gmethvin commented:
@schmitch We can already do that using `result.as("application/x-foo")`. Anything on `Result` should be accessible in a filter.

gmethvin commented:
I updated the PR with a few tests. It's worth mentioning that the cookies from the `Result` are added before cookies added with `Response#setCookie`. I don't think it makes sense to use both methods within the same action, but I added a test just to make clear what the behavior should be.

gmethvin commented:
Are there any objections to merging this as-is? I could create another PR to handle the session/flash stuff.

Speaking of that, one solution is to make the following actually mutate the session:
```scala
Result res = ok();
res.session().put("foo", "bar");
return res;
```
Since Session is designed to be mutated, it actually seems weird to me that it doesn't work like that now.

gmethvin commented:
2.5.x: 79a0b89
